### PR TITLE
Improved performance of `mars.learn.metrics.{roc_curve, roc_auc_score}`

### PR DESCRIPTION
**Pull request by qinxuye**
## What do these changes do?

This PR improved performance of `mars.learn.metrics.{roc_curve, roc_auc_score}`

### Details

- [x] Support `tileable.cache = True` to cache intermediate tileables.
- [x] Warning when duplicated execution is submitted by enabling `options.warn_duplicated_execution`.
- [x] Add ASV benchmark for `roc_curve` & `roc_auc_score`.
- [x] Improve `roc_curve` & `roc_auc_score` via eliminating duplicated execution.

#### 1. Cache intermediate tileables.

Some tileables were executed multiple times before, now we try to cache them.

#### 2. Optimize by implementing differently

Some optimization are applied including:

##### Perform fancy indexing only once

e.g. 

```
t1[ind], t2[ind]
```

now, we do

```
mt.stack([t1, t2])[:, ind]
```

##### Use DataFrame sort to replace tensor argsort + fancy indexes

e.g. 

```
ind = mt.argsort(t2)[::-1]
t1[ind], t2[ind]
```

into

```
df = md.DataFrame({0: t1, 1: t2})
df = df.sort_values(t2, ascending=False)
df[0].to_tensor(), df[1].to_tensor()
```

### Table

| | Master | Improvement 1 |  Improvement 2 |
| --- | --- | --- | --- |
| roc_curve (1 chunk) | 3.04±0.03s | 2.65±0.08s | 2.59±0.01s |
| roc_curve (n chunk) | 18.6±0.05s | 14.9±0.06s | 6.77±0.07s |
| roc_auc_score (1 chunk) | 3.28±0.01s | 2.49±0.01s | 2.43±0.02s |
| roc_auc_score (n chunk) | 17.4±0.1s | 13.2±0s | 6.19±0.01s |

## Related issue number

Fixes #xxxx

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/tests/test_session.py [hunk 1]
```diff
@@ -481,3 +482,22 @@ def test_align_series(setup):
...
+    t = mt.tensor(raw)
+    t.cache = True
+    t2 = t + 1
+    result = t2.execute().fetch()
+    np.testing.assert_array_equal(result, raw + 1)
+    np.testing.assert_array_equal(t.fetch(), raw)
+
+    t = mt.tensor(raw)
+    with option_context({"warn_duplicated_execution": True}):
+        with pytest.warns(
```
Comment (wjsi) [inline mars/tests/test_session.py:498]: Need tests for `t + 1` not producing warnings.

File: mars/deploy/oscar/session.py [hunk 2]
```diff
@@ -669,21 +675,29 @@ class ChunkFetchInfo:
...
-    session: "AbstractSession", result_tileables: List[TileableType]
-):
+    session: "AbstractSession",
+    result_tileables: List[TileableType],
+    warn_duplicated_execution: bool = False,
+) -> Tuple[TileableGraph, List[TileableType]]:
     tileable_to_copied = dict()
-    result = [None] * len(result_tileables)
+    indexer = itertools.count()
+    result_to_index = {t: next(indexer) for t in result_tileables}
```
Comment (wjsi) [inline mars/deploy/oscar/session.py:689]:
```suggestion
    result_to_index = {t: i for t, i in zip(result_tileables, indexer)}
```

**Review by wjsi — APPROVED** [review]
LGTM

**Review by hekaisheng — APPROVED** [review]
LGTM

**Pull request merged**